internal/provider/kubernetes/controller.go

```diff
@@ -1432,30 +1458,40 @@ func (r *gatewayAPIReconciler) processParamsRef(ctx context.Context, gc *gwapiv1
 		return nil
 	}
 
-	found := false
-	valid := false
-	var validationErr error
 	for i := range epList.Items {
 		ep := epList.Items[i]
 		r.log.Info("processing envoyproxy", "namespace", ep.Namespace, "name", ep.Name)
 		if classRefsEnvoyProxy(gc, &ep) {
-			found = true
+			if !gc.DeletionTimestamp.IsZero() {
+				if err := r.removeFinalizer(ctx, &ep); err != nil {
+					r.log.Error(err, fmt.Sprintf("failed to remove finalizer from envoy proxy %s",
+						ep.Name))
+					return err
+				}
+				return nil
+			}
 			if err := validation.ValidateEnvoyProxy(&ep); err != nil {
-				validationErr = fmt.Errorf("invalid envoyproxy: %v", err)
-				continue
+				return fmt.Errorf("invalid gatewayclass %s: %v", gc.Name, err)
 			}
-			valid = true
+
+			if err := r.addFinalizer(ctx, &ep); err != nil {
+				r.log.Error(err, fmt.Sprintf("failed adding finalizer to envoy proxy %s",
+					ep.Name))
+				return err
+			}
+
 			resourceTree.EnvoyProxy = &ep
 			break
 		}
-	}
 
-	if !found {
```

Comment:
why have these been removed ?
these help build status messages that surface in the GatewayClass

Reply:
Right, I thought this change was needed and wanted to discuss it.
I have added it again and tested it out, it seems I was wrong earlier.
Thanks for catching this!